### ISSUE TITLE
Under 0.5.0, output files are missing

### ISSUE THREAD

**Issue by scriby** [kotas/gulp-tsc#17]
I upgraded to 0.5.0 from 0.0.4, and I don't see any .js files after doing a build.

Here is my gulp task:

```
gulp.task('typescript', ['clean'], function() {
    var typescript = require('gulp-tsc');

    return gulp.src(['./**/*.ts', '!./node_modules/**'], { read: false })
      .pipe(typescript({ noImplicitAny: true, target: 'ES5', module: 'COMMONJS', sourcemap: true }))
      .pipe(gulp.dest('.'));
  });

 gulp.task('build', ['clean', 'typescript'], function(){

  });
```

Here is a screencast which attempts to demonstrate that I am not crazy:

http://screencast.com/t/JX1y89rA

**Comment by kotas** [comment]
Hmm... I can't reproduce your problem: https://github.com/kotas/gulp-tsc-issue-17

I have no idea what causes the problem, but "clean" task looks suspicious to me.

Do you return gulp stream in "clean" task so that gulp waits for finishing asynchronous clean-up?

```
gulp.task('clean', function () {
   return gulp.src(...).pipe(clean());
   // "return" is needed
});
```

Or how about if removing "clean" task dependency from "typescript" task and run "typescript" task alone?

---

If downgrading to v0.4.0 fixes the problem, it is obviously by changes in v0.5.0.

v0.5.0 also includes the update of depending modules, so perhaps they could have the problem...

Anyway, if you can, please try to minimize the problem. Thank you!

**Comment by scriby** [comment]
Here's a link to a minimal example:

https://drive.google.com/file/d/0BwHuMnbmJYqeenVFbmtCYXhINjQ/edit?usp=sharing

```
npm install
./node_modules/.bin/gulp clean
./node_modules/.bin/gulp typescript
```

It looks like the key is that there's no .ts file at the root. It appears to be moving everything "up" a directory. I think this is what was happening to me earlier, I just didn't notice in the context of a larger project.

**Comment by scriby** [comment]
Have you had a chance to look into this? I'd like to upgrade to the latest version to get another bugfix, but this problem is preventing me from doing so.

**Comment by kotas** [comment]
Sorry for late I've been busy this week. And thank you so much for your coop!

I will work on this soon.

**Comment by kotas** [comment]
I've investigated your minimal example.

The problem is, your source files are stored in **one** directory named `directory`. I'll explain.

For your example project, gulp-tsc runs following tsc command:

```
tsc --outDir . directory/test.ts directory/directory2/test2.ts
```

It generates `./test.js` and `./directory2/test2.js` (not `./directory/test.js` and `./directory/directory2/test2.js`).

New version of gulp-tsc follows that rule strictly due to change of how to capture generated files by tsc. I didn't notice about the incompatibility which should have changed the major version of gulp-tsc.

The reason why it happens is `tsc` skips the "common part" of input paths and uses "different part of paths" for output path.

Let's say if we run tsc command:

```
tsc --outDir out /foo/bar/src1.ts /foo/bar/baz/src2.ts
```

tsc tries to find the "common part" from those input paths. The common part is `/foo/bar/`, so tsc skips that path and uses the rest part `src1.ts` and `baz/src2.ts` for output path. Therefore tsc generates `out/src1.js` and `out/baz/src2.js`.

For relative paths, tsc first absolutize input paths and does the same thing. That's why `directory` is skipped in output path.

---

To fix the problem, I think there are two options:

a. Fix your `gulp.dest('.')` to `gulp.dest('./directory')` to include the skipped common part of input files.

b. Add gulp-tsc option to do the same thing as the option `a` automatically so that you don't have to change your `gulp.dest`.

I think it would be good to have the new option, but am worried that it is a bit complicated and can confuse users of gulp-tsc.

**Comment by scriby** [comment]
I use the same gulp task to compile several different projects with different structures. I would have to restructure things, and then keep track of the first folder that contains a .ts file, which is kind of a pain.

The funny thing is that the old version of gulp-tsc works in this scenario (sorry, I know I had something to do with you changing it).

My vote would be for gulp-tsc to handle this transparently if possible. I don't think it's uncommon to not have any source files at the top level, but nested under app, test, etc. directories.

**Comment by kotas** [comment]
All right. I think it was a mistake to change behaviors without major version up. So I'll try to simulate the old behavior by default. :smiley:

**Comment by kotas** [comment]
@scriby I have published v0.5.1 for fixing this issue. Confirmed to work with your minimal example.

Please check if it works for your projects. Thank you!

**Comment by kotas** [comment]
v0.5.1 had a bug, which is fixed in v0.6.0. So please test with v0.6.0 if you have not tested the fix yet.

I'll close this issue for now. Please re-open it if you still have a problem. Thank you!
